server/graphql/utils/requestGroup.ts

```diff
@@ -1,13 +1,43 @@
-import { ServerResponseInterface } from '../serverResponse'
+import mongoose, { Document } from 'mongoose'
+import { softDeleteRequestTypeHelper } from './requestType'
 
-const updateRequestGroupHelper = (requestGroup, dataSources): Promise<ServerResponseInterface> => {
-  return dataSources.requestGroups.update(requestGroup)
-    .then(res => {
-      return res
-    })
-    .catch(error => {
-      return error
+const updateRequestGroupHelper = async (requestGroup, dataSources): Promise<Document> => {
+  const session = await mongoose.startSession()
+  try {
+    session.startTransaction()
+    const res = await dataSources.requestGroups.update(requestGroup, session)
+    await session.commitTransaction()
+    return res
+  }
+  catch(error) {
+    console.log(error)
+    await session.abortTransaction()
+  }
+  finally {
+    session.endSession()
+  }
+}
+
+const softDeleteRequestGroupHelper = async (id, dataSources) => {
+  const session = await mongoose.startSession()
+  try {
+    session.startTransaction()
```

Comment:
Do nested sessions work as expected? Please test with erroneous input. (note that updateRequestTypeHelper creates a session and then calls this multiple times, each time creating a session)

Regardless, my guess is that multiple sessions slows down the operation.

Perhaps the session should be made in the caller of the helpers (e.g. the resolver or some wrapper function)

Reply:
It's worked with what I tested so far. I'm aware multiple sessions are created and I did this because it seemed like sessions that modified multiple collections would always throw a WriteConflict error.
I definitely agree that abstracting the session into a helper function is something we can look at going forward.

Reply:
From the documentation I read, I don't think WriteConflicts result from that. Instead, what may be happening is that you read/update the same document (e.g. request) in the same session using await. We can avoid this by either fetching the corresponding document from the cache, or by reading the current state prior to entering the session.

I'm not entirely sure, this is just a guess at why.